[Tests] Compiler tests

damirka opened the pull request:
## Motivation

Closes #916.
This PR uses @Protryon's test framework.

New framework adds significant changes to compiler tests:

1. Outputs are verified, that includes registers, state and number of constraints.
2. Tests are real world programs - they take input and must return something.
3. `console.assert`s are removed from main functions, they are prohibited within main and do not produce constraints. Test result verification is done on output level instead.
4. Manner of writing tests is now more handy and intuitive in some way. Inputs can be specified right in Leo code, and that saves time updating, debugging and writing new tests.

Please review *[testing doc](https://github.com/AleoHQ/leo/blob/compiler-tests/tests/README.md)* as well.

## Test Plan

It only consists of tests.

## Progress tracker

- [x] address
- [x] array
- [x] boolean
- [x] circuits
- [x] console
- [x] core
- [x] definition
- [x] function
- [x] import
- [x] input_files
- [x] integers
- [x] mutability
- [x] statements
- [x] tuples
- [x] global_const

## Skipped until gadget is implemented

- field
- group

## Framework improvements 

Inline input setting:
```yaml
namespace: Compile
expectation: Pass
input_file: input/dummy.in # both file and input below are going to be run
inputs: 
 - first.in: |
     [main]
     a: u32 = 100;
     
     [registers]
     r0: bool = false
 - second.in: | ...
    # use many examples as needed
```

## Found errors:

- [x] Constant tuples are failing with type mismatch: "Expected input variable `x` to be type `(u8, bool, u8)`, found type `10Unsigned(U8Type(U8Type))`"

## Skipped tests:

- pow tests on integers (both signed and un) - generates way too many constraints
- import tests - require additional work on sanitizing compiler output

acoglio reviewed (approved):
Looks good to me.

damirka commented:
@Protryon should I remove old compiler tests? Ones in `compiler/otests`.

acoglio reviewed (approved):
Looks good!

Protryon commented:
Can you add a check in the Compiler test runner for Compile namespace that asserts that the output constraint size > 0

damirka commented:
Please do not merge this PR, CIs are failing and I need to fix it, luckily now problem is known. And I’ll add constraints check with this patch.

damirka commented:
@Protryon 

Updated PR.

1. Added constraint number check, now all tests have at least one constraint or they fail.
2. Tests were failing on CI because I added CWD into compiler to enable imports tests - so diff included full path to file (including home folder): `/circle-ci/..../file.leo` vs `/Users/user/...`. 
3. Temporarily disabled compiler tests involving imports, we still have parser tests.
4. I'll implement another runner that sanitizes error output and removes full paths, but want to do it later and in another PR.

gluax reviewed (approved):
LGTM from the few files I looked at.
